### PR TITLE
add compilationinfo and related types

### PR DESCRIPTION
**Pull request by partouf**
As discussed here https://github.com/compiler-explorer/compiler-explorer/pull/4044/files#diff-ad70709cafce13a26acf40e47dced8fe40d1540f4ca54a282aa6afde39675176R140 - some types were missing.

Are now more complete-ish

### CODE REVIEW

**Review by matsjla — APPROVED** [review]

File: lib/base-compiler.ts [hunk 1]
```diff
@@ -1598,20 +1601,22 @@ export class BaseCompiler {
...
-            key,
-        );
-        compilationinfo.executableFilename = this.getExecutableFilename(
-            customBuildPath || result.dirPath,
-            this.outputFilebase,
-            key,
-        );
-        compilationinfo.asmParser = this.asm;
-        return compilationinfo;
+    getCompilationInfo(
```
Comment (mattgodbolt) [inline lib/base-compiler.ts:1604]: nice!

**Review by mattgodbolt — APPROVED** [review]

File: types/compilation/compilation.interfaces.ts [hunk 2]
```diff
@@ -121,3 +122,28 @@ export type ToolResult = {
     stdout: ResultLine[];
     artifact?: Artifact;
 };
+
+export type CompilationInfo = {
+    mtime: any;
```
Comment (RubenRBS) [inline types/compilation/compilation.interfaces.ts:127]: `lib/handle/compiler.js` in line 155 sets the `modificationTime` equal to `Stat.mtime` which is a `Date`. This also gives us a proper type for base-compiler's `mtime`, but that probably will need to be `Date | null` given that we only set it when calling `initialize`

File: types/compilation/compilation.interfaces.ts [hunk 3]
```diff
@@ -121,3 +122,28 @@ export type ToolResult = {
...
+    compiler: any;
```
Comment (RubenRBS) [inline types/compilation/compilation.interfaces.ts:128]: This compiler is base-compiler's `this.compiler`. That should at some point be partially represented by `types/compiler.interface.ts` when we're done at least passing every key in `lib/compiler-finder.js` to that interface file. Could probably set the type here if if not too many ts files complain (And extend it with Record<any, any> if they do)

**Review by RubenRBS — APPROVED** [review]

File: types/compilation/compilation.interfaces.ts (hunk 2 above)
Comment (RubenRBS) [reply types/compilation/compilation.interfaces.ts:127]: Or TIL, https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-7.html#strict-class-initialization is useful for this cases

**Comment by RubenRBS** [comment]
@partouf added most of the compiler type missing keys, let me know what you think

**Comment by partouf** [comment]
> @partouf, I've added most of the compiler type missing keys, let me know what you think

LGTM


This does make me think about 2 things;
* The type should probably be called something else than Compiler, it's too ambiguous when searching for it. Perhaps CompilerInfo?
* The client-options that's pushed to the frontend contains all of this info, it's a little crazy. Not a thing for now, but once all of these things are typescriptified we can probably separate them without breaking everything.

**Comment by RubenRBS** [comment]
Good call, I've renamed it to CompilerInfo.

As for point 2, I remember there being something to exclude some keys from the returned compilers, but I can't find anything with a quick search rn

**Pull request merged**